### ISSUE TITLE
Signature for MemoryView

### ISSUE THREAD

**Issue by KitaitiMakoto** [ruby/rbs#2215]
Hello,

When a method accepts arguments that expose MemoryView, we cannot represent that in current RBS  (v3.8.1) and type check by Steep fails. We need a way to represent an object which exposes MemoryView.

As just an idea, a special phantom module `RBS::Marker::MemoryView` might play a role in this case:

```rbs
class AMemoryViewProducer
  include RBS::Marker::MemoryView
end

class AMemoryViewConsumer
  def consume_memory_view: (RBS::Marker::MemoryView) -> void
end
```

I think we need more discussion. How do you think?

Thanks.

**Comment by KitaitiMakoto** [comment]
`RBS::Unnamed` namespace might be more proper than `Marker`.

**Comment by ksss** [comment]
Is MemoryView referring to https://docs.ruby-lang.org/en/4.0/contributing/memory_view_md.html ?
It appears to only have a C API; what are the intended use cases for it?

**Comment by KitaitiMakoto** [comment]
Yes, I mean that API.

I maintain whispercpp gem and it has RBS file: https://github.com/ggml-org/whisper.cpp/blob/master/bindings/ruby/sig/whisper.rbs

For instance, `Whisper::Context#full` accepts an object which exports a MemoryView as the second argument. I want to represent it in RBS.

It might not be suitable to use a module and `include` because it's C API as you said, but, anyway I want any way to do so.

**Comment by ksss** [comment]
I am not entirely sure about the situation, but is this link (https://github.com/ggml-org/whisper.cpp/tree/master/bindings/ruby#low-level-api-to-transcribe) the use case you have in mind? In other words, are you saying that while 'samples' cannot be defined as a specific class, you want to indicate that it is a MemoryView?

I am not very familiar with MemoryView, but if it cannot be represented by a Ruby class or module, I believe an interface is your only option.
`RBS::Unnamed` is an internal mechanism intended for ad-hoc handling of special objects like `ENV` or `ARGV`.

**Comment by KitaitiMakoto** [comment]
> while 'samples' cannot be defined as a specific class, you want to indicate that it is a MemoryView

Exactly.

Do you mean interface like Go's empty interface? MemoryView doesn't require any Ruby layer method.

**Comment by ksss** [comment]
In that situation, it is difficult to perform type checking with Ruby's type system. I think the only option is to effectively give it an `untyped` alias and aim solely for documentation purposes.

If MemoryView is implemented as a class at the Ruby level and each library provides support for typing, it might become possible to perform type checking in the long term. However, at this moment, it seems to be a limitation of Ruby's static type checking.

**Comment by KitaitiMakoto** [comment]
Thank you for the explanation. I understood the current situation.

Another question is: would you consider adding new feature to type MemoryView producers and consumers for long-term thing?

I know MemoryView is less used for now. But, Ractor-shareable objects and frozen objects share the same characteristic. Off the top of my head, how about adding `marker` feature like `interface`?

```rbs
marker _RactorShareable

class Ractor
  def self.make_shareable: [T] (T obj, ?copy: boolish) -> T & _RactorShareable
  def send: (_RactorShareable obj, ?move: boolish) -> Ractor
end
```

```rbs
marker _Frozen

class Module < Object
  # It's better if it's able to declare `self` is `extend`ed by `_Frozen` now
  def freeze: () -> self & _Frozen
end

def immutable_method: (_Frozen) -> untyped
```

```rbs
marker _MemoryViewProducer

class AudioData
  include _MemoryViewProducer
end

def process_audio: (_MemoryViewProducer) -> untyped
```

I think this idea is applicable to others like encoding-specified `String`s, sorted `Array`, open / closed `IO`, objects which have some value of instance variable, and so on.

If you interested in this type of discussion, I'd like to continue. But If not, I will close this issue. I do not insist on near-term implementation. This is intended as a starting point for discussion. And, I don't mind at all even if you're not interested.

Thank you.

**Comment by ksss** [comment]
Are you thinking in terms of the **Refinement Type** or  the **Dependent Type**?
I do not fully understand your intent, but locally, The Branded types in TypeScript might be a hint.

```rbs
# sample.rbs
type user_id = Integer & { __brand: "user_id" }
type team_id = Integer & { __brand: "team_id" }
```

```rb
class Team
  #: (user_id) -> void
  def create_by_user_id(id)
    id
  end
end

user_id = 123 #: user_id
team_id = 123 #: team_id
Team.new.create_by_user_id(123) # => Ruby::ArgumentTypeMismatch
Team.new.create_by_user_id(user_id) #=> Pass
Team.new.create_by_user_id(team_id) # => Ruby::ArgumentTypeMismatch
```

**Comment by KitaitiMakoto** [comment]
Yes, I'm thinking about a refinement-type-like feature for whispercpp gem. The branded type works perfectly for individual instances:

```rbs
type memory_view_producer = Object & { __brand: "memory_view_producer" }

class Samples
end

module Whisper
  class Context
    def full: (Whisper::Params, memory_view_producer, ?Integer n_samples) -> self
  end
end
```

```ruby
require "whisper"

class Samples
end

whisper = Whisper::Context.new("base")
params = Whisper::Params.new
samples = Samples.new #: memory_view_producer

whisper.full(params, samples)
```

However, MemoryView is registered on a class, so ideally I would like to type the class as a MemoryView producer, and then use that type in method signatures that accept MemoryView producers. I understand that this is not possible in current RBS. For the longer term, would you be open to considering a feature that makes this possible? Or, should we type using branded types?
